include/clap/ext/draft/context-menu.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include "../../plugin.h"
+
+// This extension lets the host and plugin exchange menu items and let the plugin ask the host to
+// show its context menu.
+
+static CLAP_CONSTEXPR const char CLAP_EXT_CONTEXT_MENU[] = "clap.context-menu.draft/0";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// There can be different target kind for a context menu
+enum {
+   CLAP_CONTEXT_MENU_TARGET_KIND_GLOBAL,
+   CLAP_CONTEXT_MENU_TARGET_KIND_PARAM,
+};
+
+// Describes the context menu target
+typedef struct clap_context_menu_target
+{
+   uint32_t kind;
+   clap_id id;
+} clap_context_menu_target_t;
+
+// Flags for the context menu accelerator
+enum {
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_SHIFT = (1 << 0),
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_CONTROL = (1 << 1),
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_ALT = (1 << 2),
+
+   // Windows key or Command key
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_META1 = (1 << 3),
+};
+
+// Describes a single key strike with its modifiers
+typedef struct clap_context_menu_accelerator_key {
+   uint32_t flags;
+   uint32_t key; // ASCII key
+} clap_context_menu_accelerator_key_t;
+
+// Describes an accelerator which can consist of a key strike combo
+typedef struct clap_context_menu_accelerator {
+   uint32_t combo_size;
+   clap_context_menu_accelerator_key_t keys[4];
+} clap_context_menu_accelerator_t;
+
+// Context menu builder
+typedef struct clap_context_menu_builder {
+   void *ctx;
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_entry)(const struct clap_context_menu_builder *builder,
+                             const char                             *label,
+                             const clap_context_menu_accelerator_t  *accelerator,
+                             bool                                    is_enabled,
+                             clap_id                                 action_id);
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_checkbox)(const struct clap_context_menu_builder *builder,
+                                const char                             *label,
+                                const clap_context_menu_accelerator_t  *accelerator,
+                                bool                                    is_enabled,
+                                bool                                    is_checked,
+                                clap_id                                 action_id);
+
```

Comment:
Maybe there could be way to handle redundant menu entries. E.g. if both the host and the plugin want to add a 'reset to default value' menu entry? So I think it might be a useful addition if add_entry and add_checkbox had an additional parameter which is an enum value for standard menu entries. This way the menu builder can omit redundant entries.
Useful enum values might be CUSTOM (for non-standard), RESET_TO_DEFAULT, SET_TO_MIN, SET_TO_MAX, COPY_VALUE, PASTE_VALUE, COPY_PARAM_ID_TO_CLIPBOARD, ...
If we want to go totally crazy, even submenus might be standardized (like a common 'add modulation from source' menu). So there could be a second enum for standard submenus and an additional parameter for begin_submenu.

Reply:
I think it is too specific.
Maybe by convention, plugin should only inject action that the host can't do, and vice-versa.

If the host is showing the menu and asking the plugin to populate its items, then all the operations like reset to defaults, etc... should already be provided by the host. In that case the plugin should only provide menu entries that complement what the host could achieve without this menu extension.

Reply:
So we have this problem in surge where both surge and reaper have a midi learn option and both appear on the menu. They do different things

I agree that enumerating functions is way too specific - as evidenced by the one where we have a problem not being on your list :) - and plugins just deal with this using the information provided about the host and items if it is a problem